"panic: avg: invalid partial" during vector runtime query of CSUP

## tl;dr

Using CSUP data generated from the [`hits.parquet`](https://datasets.clickhouse.com/hits_compatible/hits.parquet) data set from [ClickBench](https://github.com/ClickHouse/ClickBench):

```
$ SUPER_VAM=1 super -c "from 'hits.csup'
| url_len:=len(URL)
| summarize l:=avg(url_len), c:=count() where URL != '' by CounterID"
panic: avg: invalid partial
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:26 +0x5e
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/super/runtime/sam/op/catcher.go:25 +0x3d
panic({0xf792140?, 0xfaaf6a0?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:785 +0x132
github.com/brimdata/super/runtime/vam/expr/agg.(*avg).ConsumeAsPartial(0xc0023981c0, {0xfacf278, 0xc00275f9b0})
	/Users/phil/work/super/runtime/vam/expr/agg/avg.go:44 +0x278
github.com/brimdata/super/runtime/vam/op/summarize.(*superTable).update(0xc00249a0f0, {0xc00275f110, 0x1, 0x3}, {0xc00275f120, 0x2, 0xc00275ff80?})
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:68 +0x626
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).consume(0xc0006bd1e0, {0xc00275f110, 0x1, 0x3}, {0xc00275f120, 0x2, 0x2})
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:102 +0x4cc
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull.func1({0xc00275f110, 0x3, 0xc002b71698?})
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:83 +0x78
github.com/brimdata/super/vector.Apply(0x0?, 0xfacebd0?, {0xc00275f110?, 0xc001007828?, 0xf8c8d60?})
	/Users/phil/work/super/vector/apply.go:17 +0xa3
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull(0xc0006bd1e0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:82 +0x13b
github.com/brimdata/super/runtime/vam.(*Materializer).Pull(0xc000515c50?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:25 +0x27
github.com/brimdata/super/runtime/sam/op.(*Single).Pull(0xc0012f7140, 0x8?)
	/Users/phil/work/super/runtime/sam/op/mux.go:120 +0x33
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull(0xc0012f7380?, 0x8?)
	/Users/phil/work/super/runtime/sam/op/catcher.go:28 +0x5c
github.com/brimdata/super/runtime/exec.(*Query).Pull(0xdc8a39f?, 0x40?)
	/Users/phil/work/super/runtime/exec/query.go:49 +0x3e
github.com/brimdata/super/zbuf.CopyMux(0xc002b71d68, {0xfab50e0, 0xc0012f7200})
	/Users/phil/work/super/zbuf/mux.go:40 +0x38
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00042f000, {0xc00011a680, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:167 +0x987
github.com/brimdata/super/pkg/charm.path.run({0xc000126710, 0x1, 0x1}, {0xc00011a680, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7f
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0x10b4bfc0, {0xc00011a670, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:41 +0x5b
```

## Details

Repro is with super commit 05fae1b. The query is a simplification of the [ClickBench q27 query](https://github.com/ClickHouse/ClickBench/blob/acdd156c11043d45e9e54f791747bf51e9564d06/duckdb/queries.sql#L28). I've tried to repro using a subset of the original, full-sized data set, but have not been able.

First, download the [`hits.parquet`data set](https://datasets.clickhouse.com/hits_compatible/hits.parquet) from ClickBench.

The problem doesn't occur if I query the original Parquet in vector runtime.

```
$ super -version
Version: v1.18.0-273-g05fae1b6

$ SUPER_VAM=1 super -c "from 'hits.parquet'
| url_len:=len(URL)
| summarize l:=avg(url_len), c:=count() where URL != '' by CounterID"
{CounterID:76686(int32),l:50.352505815182916,c:4728(uint64)}
{CounterID:135730(int32),l:39.,c:3(uint64)}
...
{CounterID:215447(int32),l:52.73302155347702,c:4918(uint64)}
```

However, if I convert the Parquet data to CSUP and run the same query in vector runtime, now I see the crash.

```
$ super -f csup -o hits.csup hits.parquet

$ SUPER_VAM=1 super -c "from 'hits.csup'
| url_len:=len(URL)
| summarize l:=avg(url_len), c:=count() where URL != '' by CounterID"
panic: avg: invalid partial
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:26 +0x5e
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/super/runtime/sam/op/catcher.go:25 +0x3d
panic({0x1285d140?, 0x12b7a6a0?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:785 +0x132
github.com/brimdata/super/runtime/vam/expr/agg.(*avg).ConsumeAsPartial(0xc012c66bf0, {0x12b9a278, 0xc002a92210})
	/Users/phil/work/super/runtime/vam/expr/agg/avg.go:44 +0x278
github.com/brimdata/super/runtime/vam/op/summarize.(*superTable).update(0xc00221d6d0, {0xc0026a3ef0, 0x1, 0x3}, {0xc0026a3f00, 0x2, 0xc0026a3f80?})
	/Users/phil/work/super/runtime/vam/op/summarize/agg.go:68 +0x626
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).consume(0xc0005fd1e0, {0xc0026a3ef0, 0x1, 0x3}, {0xc0026a3f00, 0x2, 0x2})
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:102 +0x4cc
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull.func1({0xc0026a3ef0, 0x3, 0xc000c5f698?})
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:83 +0x78
github.com/brimdata/super/vector.Apply(0xf0?, 0x12b99bd0?, {0xc0026a3ef0?, 0xc00074d828?, 0x12993d60?})
	/Users/phil/work/super/vector/apply.go:17 +0xa3
github.com/brimdata/super/runtime/vam/op/summarize.(*Summarize).Pull(0xc0005fd1e0, 0x0?)
	/Users/phil/work/super/runtime/vam/op/summarize/summarize.go:82 +0x13b
github.com/brimdata/super/runtime/vam.(*Materializer).Pull(0xc0004e1338?, 0x0?)
	/Users/phil/work/super/runtime/vam/materialize.go:25 +0x27
github.com/brimdata/super/runtime/sam/op.(*Single).Pull(0xc002227320, 0x90?)
	/Users/phil/work/super/runtime/sam/op/mux.go:120 +0x33
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull(0xc0022273b0?, 0x8?)
	/Users/phil/work/super/runtime/sam/op/catcher.go:28 +0x5c
github.com/brimdata/super/runtime/exec.(*Query).Pull(0x10d5539f?, 0x40?)
	/Users/phil/work/super/runtime/exec/query.go:49 +0x3e
github.com/brimdata/super/zbuf.CopyMux(0xc000c5fd68, {0x12b800e0, 0xc002227350})
	/Users/phil/work/super/zbuf/mux.go:40 +0x38
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00040d000, {0xc00003e6b0, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:167 +0x987
github.com/brimdata/super/pkg/charm.path.run({0xc00007a700, 0x1, 0x1}, {0xc00003e6b0, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7f
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0x13c16fc0, {0xc00003e6a0, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:41 +0x5b
```

--- a/runtime/vam/op/summarize/agg.go
+++ b/runtime/vam/op/summarize/agg.go
@@ -39,13 +39,14 @@ type aggRow struct {
 func (s *superTable) update(keys []vector.Any, args []vector.Any) {
 	m := make(map[string][]uint32)
 	if len(keys) > 0 {
-		var keyBytes []byte
+		var b zcode.Builder
 		for slot := range keys[0].Len() {
-			keyBytes = keyBytes[:0]
+			b.Truncate()
 			for _, key := range keys {
-				keyBytes = key.AppendKey(keyBytes, slot)
+				key.Serialize(&b, slot)
 			}
-			m[string(keyBytes)] = append(m[string(keyBytes)], slot)
+			keyStr := string(b.Bytes())
+			m[keyStr] = append(m[keyStr], slot)
 		}
 	} else {
 		m[""] = nil

--- a/vector/any.go
+++ b/vector/any.go
@@ -9,7 +9,6 @@ type Any interface {
 	Type() super.Type
 	Len() uint32
 	Serialize(*zcode.Builder, uint32)
-	AppendKey([]byte, uint32) []byte
 }
 
 type Promotable interface {

--- a/vector/array.go
+++ b/vector/array.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -39,19 +37,6 @@ func (a *Array) Serialize(b *zcode.Builder, slot uint32) {
 		a.Values.Serialize(b, off)
 	}
 	b.EndContainer()
-}
-
-func (a *Array) AppendKey(b []byte, slot uint32) []byte {
-	b = binary.NativeEndian.AppendUint64(b, uint64(a.Typ.ID()))
-	if a.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	off := a.Offsets[slot]
-	for end := a.Offsets[slot+1]; off < end; off++ {
-		b = append(b, 0)
-		b = a.Values.AppendKey(b, off)
-	}
-	return b
 }
 
 func ContainerOffset(val Any, slot uint32) (uint32, uint32, bool) {

--- a/vector/bool.go
+++ b/vector/bool.go
@@ -66,14 +66,6 @@ func (b *Bool) Serialize(builder *zcode.Builder, slot uint32) {
 	}
 }
 
-func (b *Bool) AppendKey(bytes []byte, slot uint32) []byte {
-	var v byte
-	if b.Value(slot) {
-		v = 1
-	}
-	return append(bytes, v)
-}
-
 func (b *Bool) TrueCount() uint32 {
 	if b == nil {
 		return 0

--- a/vector/bytes.go
+++ b/vector/bytes.go
@@ -45,13 +45,6 @@ func (b *Bytes) Value(slot uint32) []byte {
 	return b.Bytes[b.Offs[slot]:b.Offs[slot+1]]
 }
 
-func (b *Bytes) AppendKey(bytes []byte, slot uint32) []byte {
-	if b.Nulls.Value(slot) {
-		return append(bytes, 0)
-	}
-	return append(bytes, b.Bytes[b.Offs[slot]:b.Offs[slot+1]]...)
-}
-
 func BytesValue(val Any, slot uint32) ([]byte, bool) {
 	switch val := val.(type) {
 	case *Bytes:

--- a/vector/const.go
+++ b/vector/const.go
@@ -45,13 +45,6 @@ func (c *Const) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (c *Const) AppendKey(bytes []byte, slot uint32) []byte {
-	if c.Nulls.Value(slot) {
-		return append(bytes, 0)
-	}
-	return append(bytes, c.val.Bytes()...)
-}
-
 func (c *Const) AsBytes() ([]byte, bool) {
 	return c.val.Bytes(), c.val.Type().ID() == super.IDBytes
 }

--- a/vector/dict.go
+++ b/vector/dict.go
@@ -29,13 +29,6 @@ func (d *Dict) Serialize(builder *zcode.Builder, slot uint32) {
 	}
 }
 
-func (d *Dict) AppendKey(bytes []byte, slot uint32) []byte {
-	if d.Nulls.Value(slot) {
-		return append(bytes, 0)
-	}
-	return d.Any.AppendKey(bytes, uint32(d.Index[slot]))
-}
-
 // RebuildDropIndex rebuilds the dictionary Index, Count and Nulls values with
 // the passed in tags removed.
 func (d *Dict) RebuildDropTags(tags ...uint32) ([]byte, []uint32, *Bool, []uint32) {

--- a/vector/dynamic.go
+++ b/vector/dynamic.go
@@ -45,7 +45,3 @@ func (d *Dynamic) Len() uint32 {
 func (d *Dynamic) Serialize(b *zcode.Builder, slot uint32) {
 	d.Values[d.Tags[slot]].Serialize(b, d.TagMap.Forward[slot])
 }
-
-func (d *Dynamic) AppendKey(b []byte, slot uint32) []byte {
-	return d.Values[d.Tags[slot]].AppendKey(b, d.TagMap.Forward[slot])
-}

--- a/vector/error.go
+++ b/vector/error.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -36,14 +34,6 @@ func (e *Error) Serialize(b *zcode.Builder, slot uint32) {
 	}
 	e.Vals.Serialize(b, slot)
 
-}
-
-func (e *Error) AppendKey(bytes []byte, slot uint32) []byte {
-	bytes = binary.NativeEndian.AppendUint32(bytes, uint32(e.Typ.ID()))
-	if e.Nulls.Value(slot) {
-		return append(bytes, 0)
-	}
-	return e.Vals.AppendKey(bytes, slot)
 }
 
 func NewStringError(zctx *super.Context, msg string, len uint32) *Error {

--- a/vector/float.go
+++ b/vector/float.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"math"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -54,21 +52,6 @@ func (f *Float) Serialize(b *zcode.Builder, slot uint32) {
 	default:
 		panic(f.Typ)
 	}
-}
-
-func (f *Float) AppendKey(b []byte, slot uint32) []byte {
-	if f.Nulls.Value(slot) {
-		b = append(b, 0)
-	}
-	val := math.Float64bits(f.Values[slot])
-	b = append(b, byte(val>>(8*7)))
-	b = append(b, byte(val>>(8*6)))
-	b = append(b, byte(val>>(8*5)))
-	b = append(b, byte(val>>(8*4)))
-	b = append(b, byte(val>>(8*3)))
-	b = append(b, byte(val>>(8*2)))
-	b = append(b, byte(val>>(8*1)))
-	return append(b, byte(val>>(8*0)))
 }
 
 func FloatValue(vec Any, slot uint32) (float64, bool) {

--- a/vector/int.go
+++ b/vector/int.go
@@ -46,21 +46,6 @@ func (i *Int) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (i *Int) AppendKey(b []byte, slot uint32) []byte {
-	if i.Nulls.Value(slot) {
-		b = append(b, 0)
-	}
-	val := i.Values[slot]
-	b = append(b, byte(val>>(8*7)))
-	b = append(b, byte(val>>(8*6)))
-	b = append(b, byte(val>>(8*5)))
-	b = append(b, byte(val>>(8*4)))
-	b = append(b, byte(val>>(8*3)))
-	b = append(b, byte(val>>(8*2)))
-	b = append(b, byte(val>>(8*1)))
-	return append(b, byte(val>>(8*0)))
-}
-
 func (i *Int) Promote(typ super.Type) Promotable {
 	return &Int{typ, i.Values, i.Nulls}
 }

--- a/vector/ip.go
+++ b/vector/ip.go
@@ -34,13 +34,6 @@ func (i *IP) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (i *IP) AppendKey(b []byte, slot uint32) []byte {
-	if i.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	return super.AppendIP(b, i.Values[slot])
-}
-
 func IPValue(val Any, slot uint32) (netip.Addr, bool) {
 	switch val := val.(type) {
 	case *IP:

--- a/vector/map.go
+++ b/vector/map.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -42,18 +40,4 @@ func (m *Map) Serialize(b *zcode.Builder, slot uint32) {
 	}
 	b.TransformContainer(super.NormalizeMap)
 	b.EndContainer()
-}
-
-func (m *Map) AppendKey(b []byte, slot uint32) []byte {
-	b = binary.NativeEndian.AppendUint64(b, uint64(m.Typ.ID()))
-	if m.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	off := m.Offsets[slot]
-	for end := m.Offsets[slot+1]; off < end; off++ {
-		b = append(b, 0)
-		b = m.Keys.AppendKey(b, off)
-		b = m.Values.AppendKey(b, off)
-	}
-	return b
 }

--- a/vector/net.go
+++ b/vector/net.go
@@ -34,13 +34,6 @@ func (n *Net) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (n *Net) AppendKey(b []byte, slot uint32) []byte {
-	if n.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	return super.AppendNet(b, n.Values[slot])
-}
-
 func NetValue(val Any, slot uint32) (netip.Prefix, bool) {
 	switch val := val.(type) {
 	case *Net:

--- a/vector/record.go
+++ b/vector/record.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -38,16 +36,4 @@ func (r *Record) Serialize(b *zcode.Builder, slot uint32) {
 		f.Serialize(b, slot)
 	}
 	b.EndContainer()
-}
-
-func (r *Record) AppendKey(b []byte, slot uint32) []byte {
-	b = binary.NativeEndian.AppendUint64(b, uint64(r.Typ.ID()))
-	if r.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	for _, f := range r.Fields {
-		b = append(b, 0)
-		b = f.AppendKey(b, slot)
-	}
-	return b
 }

--- a/vector/set.go
+++ b/vector/set.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -40,17 +38,4 @@ func (s *Set) Serialize(b *zcode.Builder, slot uint32) {
 	}
 	b.TransformContainer(super.NormalizeSet)
 	b.EndContainer()
-}
-
-func (s *Set) AppendKey(b []byte, slot uint32) []byte {
-	b = binary.NativeEndian.AppendUint64(b, uint64(s.Typ.ID()))
-	if s.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	off := s.Offsets[slot]
-	for end := s.Offsets[slot+1]; off < end; off++ {
-		b = append(b, 0)
-		b = s.Values.AppendKey(b, off)
-	}
-	return b
 }

--- a/vector/string.go
+++ b/vector/string.go
@@ -46,13 +46,6 @@ func (s *String) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (s *String) AppendKey(bytes []byte, slot uint32) []byte {
-	if s.Nulls.Value(slot) {
-		return append(bytes, 0)
-	}
-	return append(bytes, s.Bytes[s.Offsets[slot]:s.Offsets[slot+1]]...)
-}
-
 func StringValue(val Any, slot uint32) (string, bool) {
 	switch val := val.(type) {
 	case *String:

--- a/vector/type.go
+++ b/vector/type.go
@@ -46,13 +46,6 @@ func (t *TypeValue) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (t *TypeValue) AppendKey(b []byte, slot uint32) []byte {
-	if t.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	return append(b, t.Value(slot)...)
-}
-
 func TypeValueValue(val Any, slot uint32) ([]byte, bool) {
 	switch val := val.(type) {
 	case *TypeValue:

--- a/vector/uint.go
+++ b/vector/uint.go
@@ -46,21 +46,6 @@ func (u *Uint) Serialize(b *zcode.Builder, slot uint32) {
 	}
 }
 
-func (u *Uint) AppendKey(b []byte, slot uint32) []byte {
-	if u.Nulls.Value(slot) {
-		b = append(b, 0)
-	}
-	val := u.Values[slot]
-	b = append(b, byte(val>>(8*7)))
-	b = append(b, byte(val>>(8*6)))
-	b = append(b, byte(val>>(8*5)))
-	b = append(b, byte(val>>(8*4)))
-	b = append(b, byte(val>>(8*3)))
-	b = append(b, byte(val>>(8*2)))
-	b = append(b, byte(val>>(8*1)))
-	return append(b, byte(val>>(8*0)))
-}
-
 func (u *Uint) Promote(typ super.Type) Promotable {
 	return &Uint{typ, u.Values, u.Nulls}
 }

--- a/vector/union.go
+++ b/vector/union.go
@@ -1,8 +1,6 @@
 package vector
 
 import (
-	"encoding/binary"
-
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/zcode"
 )
@@ -33,14 +31,6 @@ func (u *Union) Serialize(b *zcode.Builder, slot uint32) {
 	b.Append(super.EncodeInt(int64(tag)))
 	u.Dynamic.Serialize(b, slot)
 	b.EndContainer()
-}
-
-func (u *Union) AppendKey(b []byte, slot uint32) []byte {
-	b = binary.NativeEndian.AppendUint64(b, uint64(u.Typ.ID()))
-	if u.Nulls.Value(slot) {
-		return append(b, 0)
-	}
-	return u.Dynamic.AppendKey(b, slot)
 }
 
 func Deunion(vec Any) Any {

--- a/vector/view.go
+++ b/vector/view.go
@@ -109,7 +109,3 @@ func (v *View) Len() uint32 {
 func (v *View) Serialize(b *zcode.Builder, slot uint32) {
 	v.Any.Serialize(b, v.Index[slot])
 }
-
-func (v *View) AppendKey(b []byte, slot uint32) []byte {
-	return v.Any.AppendKey(b, v.Index[slot])
-}

